docs/Storage Rules.md

```diff
@@ -0,0 +1,190 @@
+
+## Paths
+
+Paths allow you to define where your files are, and what permissions should apply to them. You can have dynamic paths and static paths.
+
+Dynamic paths can be used as variables within functions.
+
+### Folder paths
+
+Folder paths
+
+``` yaml
+paths:
+  /user/:userId/files/:
+    # Rules
+```
+
+**Note the trailing slash**. This is how you define folder paths.
+
+You can use this to add permissions to listing files in a directory or downloading a `.zip` file of all the files.
+
+### File paths
+
+File paths define rules for the individual files within your storage.
+
+``` yaml
+paths:
+  /user/:userId/files/:fileId:
+    # Rules
+```
+
+Here, the `/user` and `/files` parts are static paths. The `:userId` and `:fileId` parts are dynamic paths.
+
+Here's an example path that would be validated by this rule:
+
+``` txt
+/user/1/files/image.png
+```
+
+## Rules
+
+You can specify the following rules in your `rules.yaml` file.
+
+- create: upload a file and set metadata
+- update: update the file and patch metadata
+- get: return the file
+- list: return a zip file with all the permitted files with a key starting with /custom-path/
+- delete: delete the file
+
+For simple allow/deny, you can return boolean values. This should be a simple javascript expression in a string.
+
+For any complex permissions using variables, you should use [functions](#functions)
+
+``` yaml
+paths:
+  /public:
+    list: 'return true'
+  /private:
+    list: 'return false'
+```
+
+### File tokens
+
+Files are uploaded with a token within their metadata. This will mean that files are visible without authentication, if the file token is passed in as a query parameter.
+
+You can access query parameters on the `request` object:
+
+``` yaml
+functions:
+  validToken: 'return request.query.token === resource.Metadata.token'
+```
+
+You can now use the `validToken` function to allow anyone to see the file with the correct token:
+
+``` yaml
+functions:
+  validToken: 'return request.query.token === resource.Metadata.token'
+paths:
+  /private/:fileId:
+    list: 'return false'
+    read: 'validToken()'
+```
+
+The following URL will work for a file called `image.jpg` that has a token of `04f1449d-e567-44c9-a5b8-0f21b2eaa261`:
+
+``` txt
+/storage/o/private/image.jpg?token=04f1449d-e567-44c9-a5b8-0f21b2eaa261
+```
+
+## Functions
+
+> Please note, it is not possible to call functions inside other functions
+
+Functions allow you to define permissions which can be used by any rules. You can access the `permission_variables` cookie here, as `request.auth`.
+
+For example, if you would like to allow users to access files within a folder named as their user id, you can add the following function:
+
+``` yaml
+functions:
+  isOwner: '!!request.auth && request.auth["user-id"] === userId'
+```
+
+This function uses a variable, called `userId`, which must be passed in by the rule from a [dynamic path](#folder-paths):
```

Comment:
Maybe explain the `!!request.auth` too? I guess it just makes sure that the `auth` key exists on `request` (you are logged in).

Reply:
Just to clarify, would it break anything if `!!request.auth` isn't there, and I made a request _without_ auth? I know in JS it would throw an error `Cannot read 'user-id' of undefined`, so is this handled gracefully into a false?

Reply:
Don't know. Can you test it? :D

Reply:
I think we get a `{"statusCode":400,"error":"Bad Request"}` when `!!request.auth` isn't there first.

Reply:
Sort of make sense. This is also how Firebase is handle it, it seems (https://firebase.google.com/docs/storage/security/user-security#user_private).

Reply:
I'll have another look soon, just got a few things to finish off.